packages/fx-core/src/core/middleware/consolidateLocalRemote.ts

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import {
+  assembleError,
+  AzureSolutionSettings,
+  ConfigFolderName,
+  err,
+  InputConfigsFolderName,
+  Inputs,
+  LogProvider,
+  Platform,
+  StatesFolderName,
+  TeamsAppManifest,
+} from "@microsoft/teamsfx-api";
+import { isSPFxProject, isAADEnabled, isConfigUnifyEnabled } from "../../common/tools";
+import { environmentManager } from "../environment";
+import { CoreSource, ConsolidateCanceledError } from "../error";
+import { Middleware, NextFunction } from "@feathersjs/hooks/lib";
+import fs from "fs-extra";
+import path from "path";
+import {
+  BotOptionItem,
+  MessageExtensionItem,
+  TabOptionItem,
+} from "../../plugins/solution/fx-solution/question";
+import { APP_PACKAGE_FOLDER_FOR_MULTI_ENV } from "../../plugins/resource/appstudio/constants";
+import { getLocalAppName } from "../../plugins/resource/appstudio/utils/utils";
+import {
+  Component,
+  ProjectMigratorGuideStatus,
+  ProjectMigratorStatus,
+  sendTelemetryErrorEvent,
+  sendTelemetryEvent,
+  TelemetryEvent,
+  TelemetryProperty,
+} from "../../common/telemetry";
+import { CoreHookContext } from "../types";
+import { TOOLS } from "../globalVars";
+import { getLocalizedString } from "../../common/localizeUtils";
+import { createManifest } from "../../plugins/resource/appstudio/plugin";
+import { getManifestTemplatePath } from "../../plugins/resource/appstudio/manifestTemplate";
+import { getTemplatesFolder } from "../../folder";
+import { loadProjectSettings } from "./projectSettingsLoader";
+import { addPathToGitignore, needMigrateToArmAndMultiEnv } from "./projectMigrator";
+
+const upgradeButton = "Upgrade";
+let fromReloadFlag = false;
+const backupFolder = ".backup";
+
+export const ProjectConsolidateMW: Middleware = async (
```

Comment:
are we going to upgrade the project version in `projectSettings.json`?

Reply:
For now, the project version has not been changed when creating a new project or upgrade an old project.

Reply:
To be consistent with Go Prod migration (https://github.com/OfficeDev/TeamsFx/pull/2988) , we may need do it as well, and this can also help us simplify the logic to detect whether current project needs to be migrated, e.g. only project version (2.0.0<= v < 3.0.0) needs to be upgraded.

Any thoughts? @xiaolang124 @jayzhang @kimizhu

Reply:
For incremental sprint, project settings version will not be upgraded. V3 API sprint will update project settings version.

Reply:
I see, thanks for the confirmation.